src/tree/index.tsx

```diff
@@ -0,0 +1,315 @@
+	const defaultRenderer = (n: TreeNode) => n.value;
+	const [itemRenderer] = children();
+
+	// convert TreeNode to LinkedTreeNodes
+	const nodeMap = new Map<string, LinkedTreeNode>();
+	for (let node of nodes) {
+		nodeMap.set(node.id, {
+			id: node.id,
+			node,
+			children: []
+		});
+	}
+	const rootNodes: LinkedTreeNode[] = [];
+	for (let linked of nodeMap.values()) {
+		if (linked.node.parent) {
+			const parent = nodeMap.get(linked.node.parent);
+			if (parent) {
+				parent.children.push(linked);
+			}
+		} else {
+			rootNodes.push(linked);
+		}
+	}
+
+	return (
+		<ol classes={[classes.root, theme.variant()]}>
+			{rootNodes.map((node) => (
+				<li classes={classes.child}>
+					<Node
+						level={0}
+						activeNode={icache.get('activeNode')}
+						checkable={checkable}
+						selectable={selectable}
+						checkedNodes={checkedNodes || []}
+						selectedNode={icache.get('selectedNode')}
+						disabledNodes={disabledNodes || []}
+						expandedNodes={expandedNodes || []}
+						node={node}
+						onActive={(n) => {
+							icache.set('activeNode', n);
+						}}
+						onSelect={(n) => {
+							icache.set('selectedNode', n);
+							onSelect && onSelect(n);
+						}}
+						onCheck={(n, c) => {
+							onCheck && onCheck(n, c);
+						}}
+						onExpand={(n, e) => {
+							onExpand && onExpand(n, e);
+						}}
+					>
+						{itemRenderer || defaultRenderer}
+					</Node>
+				</li>
+			))}
+		</ol>
+	);
+});
+
+/*******************
+ * TreeNode
+ *******************/
+
+interface TreeNodeProperties {
+	checkable: boolean;
+	level: number;
+	selectable: boolean;
+	activeNode?: string;
+	checkedNodes: string[];
+	selectedNode?: string;
+	disabledNodes: string[];
+	expandedNodes: string[];
+	node: LinkedTreeNode;
+	onActive(node: string): void;
+	onSelect(node: string): void;
+	onCheck(node: string, checked: boolean): void;
+	onExpand(node: string, expanded: boolean): void;
+}
+
+interface TreeNodeChildren {
+	(node: TreeNode): RenderResult;
+}
+
+interface TreeNodeCache {
+	node: TreeNode;
+	expanded: boolean;
+	checked: boolean;
+	selected: boolean;
+}
+
+const treeNodeCache = createICacheMiddleware<TreeNodeCache>();
+const treeNodeFactory = create({ theme, icache: treeNodeCache, diffProperty })
+	.properties<TreeNodeProperties>()
+	.children<TreeNodeChildren>();
+
+const Node = treeNodeFactory(function({
+	middleware: { theme, icache, diffProperty },
+	properties,
+	children
+}) {
+	const {
+		node,
+		checkable,
+		level,
+		selectable,
+		activeNode,
+		selectedNode,
+		checkedNodes,
+		disabledNodes,
+		expandedNodes,
+		onActive,
+		onSelect,
+		onCheck,
+		onExpand
+	} = properties();
+	const [itemRenderer] = children();
+
+	// Merge controlled properties with our internally-tracked ones
+	diffProperty(
+		'expandedNodes',
+		(
+			{ expandedNodes: current }: TreeNodeProperties,
+			{ expandedNodes: next }: TreeNodeProperties
+		) => {
+			if ((current && current.includes(node.id)) || (next && next.includes(node.id))) {
```

Comment:
if you're going to use `node` here you need to take it off the properties passed into `diffProperty` not the main render block, because `diffProperty` is executed at a different point (before the render) so `node` can be stale.

Reply:
Good point.